### DISCUSSION TITLE
Build Kernel or image on arm64 (opi5+)

### DISCUSSION THREAD

**Discussion by bagong** [Joshua-Riek/ubuntu-rockchip#528] [Q&A]
Hi,
I would like to make this my little learning project for building kernels and os-images. However I don't have much prior knowledge, maybe I am megalomaniacal, but then it doesn't harm much to fail. It seems so lovely to have an totally transparent build system to start from. The build system is designed for x86-machines, it cross-compiles. I guess that's because Github requires this, but maybe also because conventional build machines are more powerful. My question is: is it totally unrealistic to run these builds on an opi5+? The other day I built a kernel for rk-3588 mandriva (the build seemed to go through fine, but I didn't know how to make u-boot use it ;-) Now I am trying my luck here). It took about an hour. Not sure how much added time would be required to assemble the distro files. However the main question is: if it was realistic to use an opi5+, would I merely have to remove the "cross-compile" references in the scripts and replace them by straightforward arm64, or is the fact that it cross-compiles baked deeper into the scripts? I guess all the qemu-related dependencies serve the cross-compile purpose, but I assume it doesn't harm to have that stuff available on the system, it would simply not be used, right? So my main question is: does it only require simple modifications to the scripts? Are there maybe some tips?

**Comment by hakotani-o** [comment]
I'm new to Linux, but I enjoy building disk images with Opi5, writing them to disks, and booting them.
You can build it by following the steps explained in Building in the wiki.
I want to build quickly, so I run sudo ./build.sh --board=orangepi-5 -l -do.

**Reply by bagong** [reply]
Which means you don't actually build the kernel, right? (-l: use kernel and uboot from launchpad, -do build desktop image only). I am particularly curios of the kernel build, and wonder how compiling a new kernel interferes with u-boot. Grub provides simple cli-steps to adjust to a new kernel and allows fallback to the old kernel if the new build cannot boot or has problems. I don't need that comfort, wouldn't mind copying the old kernel back on the SD card, but I need to know how to adjust u-boot to a new kernel. I guess the workflow I am looking for (thinking while writing ;-)) is to compile a new kernel and boot from it to see the results, so not really generating an image. I haven't yet bumped into a "u-boot for dummy users" documentation, which shows how to do that. The documentation I saw seemed very technical to me ;-)

**Reply by bagong** [reply]
This is a pretty good explanation (for me ;-)):
https://bootlin.com/blog/using-device-tree-overlays-example-on-beaglebone-boards/

**Reply by hakotani-o** [reply]
If you want to build the Opi5 kernel, please run sudo ./build.sh --board=orangepi-5 -so.

**Reply by bagong** [reply]
Thanks @hakotani-o , I am trying to go a bit deeper. I have the server image running and would like to build a recompiled kernel to use within that system. I haven't found the part of the scripts yet, where the u-boot bit is executed. But I'll just have to dig deeper, I guess ;-)

**Reply by hakotani-o** [reply]
I'm glad.
Scripts is scripts/build-kernel.sh and scripts/build-u-boot.sh.
build/linux-rockchip is kernel source code.
build/u-boot-orangepi-rk3588 is u-boot source code.
good luck.

**Comment by Joshua-Riek** [comment]
You can build on the Opi5 without issue, cross-compiling does not harm the build process even when running on arm64 platform.

Since I have so many development RK3588 boards I would like to set them up as builders for GitHub when I have extra time. So it's very reasonable to build directly on your Opi5.

**Comment by bagong** [comment]
In my quest to get to use u-boot manually and learn from the scripts here, I seem to have come to the conclusion that the real secret is hidden in the debian package build command for u-boot. If I build the mainline kernel and u-boot with the build script, it tells me that kernel version is allready installed (makes sense) and u-boot will be downgraded. The later seems to indicate that the u-boot package in the ppa is newer than what you'd get using the build script? I also wonder if u-boot for linux-rockchip and mainline would be the same?
